### PR TITLE
chore: make `impl Default for StateBuilder` generic

### PR DESCRIPTION
**Pull request by DaniPopes**
Follow-up to #687, see https://github.com/bluealloy/revm/pull/687#discussion_r1316523492

### CODE REVIEW

**Review by rakita — CHANGES_REQUESTED** [review]
This is more verbose than it was. Using `EmptyDB` removes the need to specify the type for `default()`

**Comment by DaniPopes** [comment]
This is only true with `StateBuilder`, which can be changed by making the type generic default to `EmptyDB`. But I don't see why `Default` (or other traits in the future) are only implemented for the default type.

File: crates/revm/src/db/states/state_builder.rs [hunk 1]
```diff
@@ -29,26 +26,38 @@ pub struct StateBuilder<DB> {
...
-}
-
-impl<DB: Database> StateBuilder<DB> {
-    /// Create default instance of builder.
-    pub fn new() -> StateBuilder<Box<EmptyDB>> {
-        StateBuilder::<Box<EmptyDB>>::default()
-    }
 
+    #[doc(hidden)]
+    #[deprecated = "instantiate with `new_with_database` instead"]
```
Comment (rakita) [inline crates/revm/src/db/states/state_builder.rs:60]: I use this fn inside reth, maybe it is not needed as there is `fn with_database` currently in the main, but can you remove `deprecated` for now?

**Review by rakita — APPROVED** [review]
Pending conflicts, but lgtm!

I added  `State::builder()` inside State so `StateBuilder` doesn't need to be specified in usually flows. So we can do `State::builder().build()` pattern.

**Comment by DaniPopes** [comment]
Re Based

File: crates/revm/src/db/states/state_builder.rs [hunk 2]
```diff
@@ -32,27 +29,37 @@ pub struct StateBuilder<DB> {
...
-}
-
-impl<DB: Database> StateBuilder<DB> {
-    /// Create default instance of builder.
-    pub fn new() -> StateBuilder<EmptyDB> {
-        StateBuilder::default()
-    }
 
-    /// With generic database.
+    /// Set the database. Prefer using [`new_with_database`](Self::new_with_database) when possible.
```
Comment (rakita) [inline crates/revm/src/db/states/state_builder.rs:62]:
```suggestion
    /// Set the database.
```

**Pull request merged**